src/cmap/auth/mongodb_oidc.ts

```diff
@@ -11,29 +20,131 @@ export interface OIDCMechanismServerStep1 {
   accessToken: string;
   expiresInSeconds?: number;
   refreshToken?: string;
 }
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export type OIDCRequestFunction = (
   idl: OIDCMechanismServerStep1
 ) => Promise<OIDCRequestTokenResult>;
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export type OIDCRefreshFunction = (
   idl: OIDCMechanismServerStep1,
   result: OIDCRequestTokenResult
 ) => Promise<OIDCRequestTokenResult>;
+
+/** @internal */
+const kCache = Symbol('cache');
+
+export class MongoDBOIDC extends AuthProvider {
+  /** @internal */
+  [kCache]: Map<string, OIDCMechanismServerStep1>;
+
+  /**
+   * Instantiate the auth provider.
+   */
+  constructor() {
+    super();
+    this[kCache] = new Map();
+  }
+
+  /**
+   * Authenticate using OIDC
+   */
+  override auth(authContext: AuthContext, callback: Callback): void {
+    const { connection, credentials } = authContext;
+
+    if (!credentials) {
+      return callback(new MongoMissingCredentialsError('AuthContext must provide credentials.'));
+    }
+
+    saslStart(connection, credentials)
+      .then(result => {
+        return saslContinue(connection, credentials, result);
+      })
+      .then(result => {
+        callback(undefined, result);
+      })
+      .catch(error => {
+        callback(error);
+      });
+  }
+
+  /**
+   * Add the specualtive auth for the initial handshake.
+   */
+  override prepare(
+    handshakeDoc: HandshakeDocument,
+    authContext: AuthContext,
+    callback: Callback<HandshakeDocument>
+  ): void {
+    callback(undefined, handshakeDoc);
+  }
+
+  /**
+   * Clear the token cache.
+   */
+  clearCache(): void {
+    this[kCache].clear();
+  }
+}
+
+/**
+ * Execute the saslStart command.
+ */
+async function saslStart(
+  connection: Connection,
+  credentials: MongoCredentials
+): Promise<OIDCMechanismServerStep1> {
+  // TODO: We can skip saslStart if we have a DEVICE_NAME or a cached token.
+  const command = saslStartCommand(credentials);
+  const result = await connection.command(credentials.source, command, undefined);
+  console.log('saslStart result', result);
+  return { clientId: '' };
+}
+
+/**
+ * Execute the saslContinue command.
+ */
+async function saslContinue(
+  connection: Connection,
+  credentials: MongoCredentials,
+  result: OIDCMechanismServerStep1
+): Promise<OIDCRequestTokenResult> {
+  const command = saslContinueCommand(1, 'test');
+  const commandResult = await connection.command(credentials.source, command, undefined);
+  console.log('saslContinue result', commandResult);
+  return { accessToken: '' };
+}
+
+/**
+ * Generate the saslStart command document.
+ */
+function saslStartCommand(credentials: MongoCredentials): Document {
+  const payload: Document = {};
+  if (credentials.username || credentials.mechanismProperties.PRINCIPAL_NAME) {
+    payload.n = credentials.username || credentials.mechanismProperties.PRINCIPAL_NAME;
```

Comment:
Small heads up, `PRINCIPAL_NAME` got dropped from the spec text, with `username` being the relevant field now

Reply:
Uses only username now.